### DISCUSSION TITLE
Ошибка после ручной правки истории прослушиваний

### DISCUSSION THREAD

**Discussion by ilyanikolaich** [Chimildic/goofy#45] [Помощь с ошибкой] [answered]
в логе выполнения функции `updateRecentTracks_` появилась такая ошибка 
```
8 февр. 2021 г., 11:30:57	 Информация	Новых треков: 5 в файле SpotifyRecentTracks.json
8 февр. 2021 г., 11:31:01	Информация	Общее количество: 2513
8 февр. 2021 г., 11:31:04	Информация	Новых треков: 8 в файле LastfmRecentTracks.json
8 февр. 2021 г., 11:31:07	Информация	Общее количество: 4455
8 февр. 2021 г., 11:31:10	 Ошибка TypeError: Cannot read property 'played_at' of undefined
    at compareDate(Library:1428:26)
    at [unknown function](Library:1391:40)
    at sortMeta(Library:1391:25)
    at [unknown function](Library:1352:17)
    at updateBothSourceRecentTracks(Library:606:15)
    at update(Library:591:13)
    at updateRecentTracks_(Library:20:18)
```

появилась приблизительно одновременно с обновлением мной кода до версии 1.4.3
(переобновил код у себя еще раз, ошибка сохраняется)
даты мзменеий на гугль драйве в файлах истории споти и ластфм отдельно - свежие, а общего файла - осталась приблизительно с начала появления ошибки, что и логично.

**Comment by Chimildic** [comment]
Вероятно, связано с вашим ручным добавлением треков в историю. 
Выложите файл `LastfmRecentTracks`. Он в гугл диске.

**Reply by ilyanikolaich** [reply]
Прилагаю
[LastfmRecentTracks.zip](https://github.com/Chimildic/goofy/files/5941870/LastfmRecentTracks.zip)

**Comment by Chimildic** [comment, marked as answer]
Запустите следующую функцию один раз. Потом ее можно удалить. Для ручного добавления треков используйте [appendTracks](https://chimildic.github.io/goofy/#/func?id=appendtracks).
```js
function patchHistory() {
  let tracks = Cache.read('LastfmRecentTracks');
  tracks.forEach((t) => {
    if (t.hasOwnProperty('added_at')) {
      t.played_at = t.added_at;
      delete t.added_at;
    } else if (!t.hasOwnProperty('played_at')) {
      t.played_at = DEFAULT_DATE;
    }
  });
  Cache.write('LastfmRecentTracks', tracks);
}
```

**Reply by ilyanikolaich** [reply]
Починилось, огромное спасибо!
>Для ручного добавления треков используйте appendTracks.

Так конечно, когда есть штатные средства, лучше использовать их.
Но мне теперь нет необходимости обновлять файлы истории - сейчас уже всё там.
Еще раз большое спасибо!
